### PR TITLE
Doc: Add documentation on TreeLB usage

### PR DESCRIPTION
**Pull request by kavithachandrasekar**

### CODE REVIEW

File: doc/charm++/manual.rst [hunk 1]
```diff
@@ -2701,6 +2701,74 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
 
+**TreeLB and its configuration**
+
+Recent changes in the LB infrastructure add support for a user-configurable
+hierarchical load balancing using TreeLB. While the legacy centralized
+strategies above are still supported, TreeLB allows performing load
+balancing at multiple levels and supports load balancing trees with different
+levels. For example, a 2-level tree consists of PEs and a root while a
+3-level tree consists of PEs, processes and a root at the top. A 4-level
+tree consists of PEs, proceeses, ProcessGroup and a root. The load balancing
```
Comment (evan-charmworks) [inline doc/charm++/manual.rst:2712]:
```suggestion
tree consists of PEs, processes, ProcessGroup and a root. The load balancing
```

File: doc/charm++/manual.rst [hunk 2]
```diff
@@ -2701,6 +2701,74 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+strategy to be used at each level and frequency at which to invoke LB
+at each level can be specified using command line parameters or using a json
+config file. We provide examples of these below:
+
+Creating a 2-level tree that uses GreedyRefine strategy at the root
```
Comment (evan-charmworks) [inline doc/charm++/manual.rst:2717]: @matthiasdiener Is there any syntax we should use for a heading like here and below?

Comment (matthiasdiener) [reply doc/charm++/manual.rst:2717]: We could maybe add another level of heading here, but it's not likely. One of the more silly aspects of RST vs. MD is the heading levels, which are far less clear in RST. In any case we should probably create a heading for `**TreeLB and its configuration**` etc. first, which would probably be better in another PR.

File: doc/charm++/manual.rst [hunk 3]
```diff
@@ -2701,6 +2701,74 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+Recent changes in the LB infrastructure add support for a user-configurable
+hierarchical load balancing using TreeLB. While the legacy centralized
+strategies above are still supported, TreeLB allows performing load
+balancing at multiple levels and supports load balancing trees with different
+levels. For example, a 2-level tree consists of PEs and a root while a
+3-level tree consists of PEs, processes and a root at the top. A 4-level
+tree consists of PEs, processes, ProcessGroup and a root. The load balancing
+strategy to be used at each level and frequency at which to invoke LB
+at each level can be specified using command line parameters or using a json
+config file. We provide examples of these below:
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2715]: There are only examples of the json files, not the command line parameters?

File: doc/charm++/manual.rst [hunk 4]
```diff
@@ -2701,6 +2701,74 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
 
+**TreeLB and its configuration**
+
+Recent changes in the LB infrastructure add support for a user-configurable
+hierarchical load balancing using TreeLB. While the legacy centralized
+strategies above are still supported, TreeLB allows performing load
+balancing at multiple levels and supports load balancing trees with different
+levels. For example, a 2-level tree consists of PEs and a root while a
+3-level tree consists of PEs, processes and a root at the top. A 4-level
+tree consists of PEs, processes, ProcessGroup and a root. The load balancing
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2712]:
```suggestion
tree consists of PEs, processes, ProcessGroups and a root. The load balancing
```

File: doc/charm++/manual.rst [hunk 5]
```diff
@@ -2701,6 +2701,74 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+    "tree": "PE_Root",
+    "Root":
+    {
+        "pe": 0,
+        "strategies": ["GreedyRefine"]
+    }
+  }
+
+Creating a 3-level tree that uses Greedy strategy at process level
+and GreedyRefine strategy at the root
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2731]:
```suggestion
Creating a 3-level tree that uses the Greedy strategy at process level
and the GreedyRefine strategy at the root:
```

File: doc/charm++/manual.rst [hunk 6]
```diff
@@ -2701,6 +2701,74 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+        "strategies": ["GreedyRefine"]
+    },
+    "Process":
+    {
+        "strategies": ["Greedy"]
+    }
+  }
+
+Creating a 4-level tree that uses GreedyRefine strategy at process level,
+process-group level and at the root
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2750]:
```suggestion
Creating a 4-level tree that uses the GreedyRefine strategy at process level,
process-group level and at the root:
```

File: doc/charm++/manual.rst [hunk 7]
```diff
@@ -2701,6 +2701,74 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+strategies above are still supported, TreeLB allows performing load
+balancing at multiple levels and supports load balancing trees with different
+levels. For example, a 2-level tree consists of PEs and a root while a
+3-level tree consists of PEs, processes and a root at the top. A 4-level
+tree consists of PEs, processes, ProcessGroup and a root. The load balancing
+strategy to be used at each level and frequency at which to invoke LB
+at each level can be specified using command line parameters or using a json
+config file. We provide examples of these below:
+
+Creating a 2-level tree that uses GreedyRefine strategy at the root
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2717]:
```suggestion
Creating a 2-level tree that uses the GreedyRefine strategy at the root:
```

File: doc/charm++/manual.rst (hunk 3 above)
Comment (kavithachandrasekar) [reply doc/charm++/manual.rst]: I've removed the text on command line parameters since that is not supported currently.

**Review by rbuch — DISMISSED** [review]
It'd be good if others could review this thoroughly for clarity, it's a bit complicated to explain everything here, so I'd like to make sure it's as lucid as possible.

File: doc/charm++/manual.rst [hunk 8]
```diff
@@ -9792,6 +9942,47 @@ to the Charm++ build system:
...
+
+Writing a load balancing strategy with TreeLB is very simple. It involves
+implementing a class inheriting from the abstract `Strategy` class
+defined in the `TreeStrategyBase.h` header file. Specifically, the
+load balancing strategy needs to be implemented in the `solve` method,
+which receives objects and processors as vectors. The solution
+needs to be stored in the `solution` object and is used by the load
+balancing framework to perform object migrations accordingly. The
+`bool objSorted` argument specifes whether the incoming vector of
+objects is sorted by load values or not. This new strategy can written
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:9956]:
```suggestion
objects is sorted by load values or not. This new strategy can be written
```

File: doc/charm++/manual.rst [hunk 9]
```diff
@@ -2701,6 +2701,156 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+centralized load balancing), the 3-level tree consists of PEs,
+processes, and a root at the top, and the 4-level tree consists of
+PEs, processes, process groups and a root. Each level only balances
+load within its corresponding domain; for example, for the 3-level
+PE-Process-Root tree during process steps, each process runs the
+specified LB strategy over only the PEs and objects contained within
+the process, while, at root steps, the root strategy is run over all
+PEs and objects in the job. The load balancing strategy to be used at
+each level and frequency at which to invoke LB at each level can be
+specified using a JSON configuration file with name `treelb.json` or
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2720]: Should this be **\`\`treelb.json\`\`** (also in the rest of the file)?

File: doc/charm++/manual.rst [hunk 10]
```diff
@@ -9715,8 +9865,8 @@ above can be used for this.
 
 .. _lbWriteNewLB:
 
-Writing a new load balancing strategy
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Writing a new load balancing strategy (Legacy centralized load balancer)
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:9868]: Just to clarify - is writing a new strategy in the old way still possible?

Comment (matthiasdiener) [reply doc/charm++/manual.rst:9868]: Also, does the section starting on line 9916 apply to the old and/or new way?

Comment (matthiasdiener) [reply doc/charm++/manual.rst:9868]: In any case, I think we might consider removing the legacy documentation.

File: doc/charm++/manual.rst [hunk 11]
```diff
@@ -2701,6 +2701,156 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+  }
+
+Creating a 4-level tree that uses the GreedyRefine strategy at the process
+and process group levels. The number of user-specified process groups is four in
+this example. A strategy is not allowed at root level for a 4-level tree since
+communicating all object load information to the root can be expensive given
+the size of the PE tree. Load is balanced at the process group level
+every five steps and at the root level every ten steps. Note also that
+the process group usage of GreedyRefine provides a custom parameter to
+the strategy. This will only be used for the process group level version of
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2765]:
```suggestion
the strategy. This parameter will only be used for the process group level version of
```

File: doc/charm++/manual.rst [hunk 12]
```diff
@@ -2701,6 +2701,156 @@ applications. RotateLB and RandCentLB are more useful for debugging
...
+**TreeLB and its configuration**
+
+TreeLB allows for user-configurable hierarchical load balancing. While
+the legacy centralized strategies above are still supported, TreeLB
+allows load balancing to be performed at different levels and
+frequencies in a modular way. TreeLB includes several kinds of trees:
+the 2-level tree consists of PEs and a root (essentially the same as
+centralized load balancing), the 3-level tree consists of PEs,
+processes, and a root at the top, and the 4-level tree consists of
+PEs, processes, process groups and a root. Each level only balances
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:2713]: I'm not sure the meaning of "process group"  becomes clear in this section.

File: doc/charm++/manual.rst [hunk 13]
```diff
@@ -9792,6 +9942,47 @@ to the Charm++ build system:
...
+  public:
+    void solve(std::vector<O>& objs, std::vector<P>& procs, S& solution, bool objsSorted)
+    {
+      int index = 0;
+      for (const auto& o : objs)
+      {
+        P p = procs[index];
+        /// The strategy goes here
+        /// The strategy goes here
+        /// The strategy goes here
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst:9977]: Is there a way you could describe a simple strategy here (e.g. something like a RandomLB)? It's not clear which of the variables need to be modified here.

File: doc/charm++/manual.rst (hunk 10 above)
Comment (rbuch) [reply doc/charm++/manual.rst:9868]: Yes, writing a strategy in the old way is still possible. The section starting on 9916 applies to only the old way, the new one doesn't involving changing linking at all, it just adds another strategy to the set of the ones that TreeLB can plug into its structure.

It should certainly be cleaned up, the trouble is that both the old and the new way have their purposes for now: the new for anything that can fit into the structure of TreeLB, the old for legacy support and for strategies that are unsuitable for the current TreeLB (they use the comm graph, are distributed, etc.).

**Review by matthiasdiener — APPROVED** [review]

**Review by rbuch — APPROVED** [review]

**Pull request merged**